src/chartDownloader.ts

```diff
@@ -0,0 +1,431 @@
+                    }
+                } catch (err) {
+                    console.error(`Error checking disk space:`, err);
+                    break;
+                }
+            }
+            promises.push(limit(async () => {
+                if (this.cancelRequested)
+                    return;
+                const buffer = await ChartDownloader.getTileFromCacheOrRemote(this.chartsPath, this.provider, tile);
+                if (buffer === null) {
+                    this.failedTiles += 1;
+                } else {
+                    this.downloadedTiles += 1;
+                }
+            }));
+            tileCounter++
+        }
+        try {
+            await Promise.all(promises);
+
+        } catch (err) {
+            // silent failure, caller can log if needed
+            console.error(`Error downloading tiles:`, err);
+        }
+        this.status = Status.Stopped;
+    }
+
+    async deleteCache(): Promise<void> {
+        this.status = Status.Running;
+        for (const tile of this.tiles) {
+            if (this.cancelRequested) break;
+            const tilePath = path.join(this.chartsPath, `${this.provider.name}`, `${tile.z}`, `${tile.x}`, `${tile.y}.${this.provider.format}`);
+            if (fs.existsSync(tilePath)) {
+                try {
+                    fs.promises.unlink(tilePath);
+                    this.cachedTiles -= 1;
+                    this.cachedTiles = Math.max(this.cachedTiles, 0);
+                } catch (err) {
+                    console.error(`Error deleting cached tile ${tilePath}:`, err);
+                }
+            }
+        }
+        this.status = Status.Stopped;
+    }
+
+    public cancelJob() {
+        this.cancelRequested = true;
+    }
+
+    private filterCachedTiles(allTiles: Tile[]): Tile[] {
+        const uncachedTiles: Tile[] = [];
+        for (const tile of allTiles) {
+            const tilePath = path.join(this.chartsPath, `${this.provider.name}`, `${tile.z}`, `${tile.x}`, `${tile.y}.${this.provider.format}`);
+            if (!fs.existsSync(tilePath)) {
+                uncachedTiles.push(tile);
+            }
+        }
+        return uncachedTiles;
+    }
+
+    public info() {
+        return {
+            id: this.id,
+            chartName: this.provider.name,
+            regionName: this.areaDescription,
+            totalTiles: this.totalTiles,
+            downloadedTiles: this.downloadedTiles,
+            cachedTiles: this.cachedTiles,
+            failedTiles: this.failedTiles,
+            progress: this.totalTiles > 0 ? (this.downloadedTiles + this.cachedTiles + this.failedTiles) / this.totalTiles : 0,
+            status: this.status
+        };
+    }
+
+    static async getTileFromCacheOrRemote(chartsPath: string, provider: any, tile: Tile): Promise<Buffer | null> {
+        const tilePath = path.join(chartsPath, `${provider.name}`, `${tile.z}`, `${tile.x}`, `${tile.y}.${provider.format}`);
+        if (fs.existsSync(tilePath)) {
```

Comment:
I strongly advice against using the Sync versions of fs methods. Afaik they completely stop the main event loop, whereas awaiting promises allow other things to proceed when waiting for i/o.

Another point is that checking for a file and then using it is an antipattern, even if here we don't expect files to disappear, but just try to read it:
> The Node.js community strongly advises against checking for a file's existence before attempting an operation, primarily because of race conditions.
> If a file exists when you call fs.exists, another process could delete it right before you call fs.readFile, causing your program to crash (or throw an error anyway).
> The modern, idiomatic Node.js solution is to use fs.promises.access and simply try the operation and handle the error